Make external event payloads strongly typed

tnotheis opened the pull request:
# Readiness checklist

-   [ ] I added/updated unit tests.
-   [ ] I added/updated integration tests.
-   [x] I ensured that the PR title is good enough for the changelog.
-   [x] I labeled the PR.

tnotheis commented on Modules/Synchronization/src/Synchronization.Domain/Entities/Sync/ExternalEvent.cs:
```diff
@@ -44,6 +43,11 @@ public ExternalEvent(ExternalEventType type, IdentityAddress owner, long index,
     public SyncRunId? SyncRunId { get; private set; }
     public IReadOnlyCollection<SyncError> Errors => _errors;
 
+    public void UpdateIndex(long newIndex)
+    {
+        Index = newIndex;
+    }
+
```
Hint for the reviewer: this method is necessary now, because we cannot recreate the External Event in the SynchronizationDbContext anymore (because the generic SynchronizationDbContext.CreateExternalEvent method doesn't know which external event to create).

But even though letting the Index default to `0` in the cosntructor is kind of ugly, I still think this apporach is somewhat better than the old one, because now the SynchronizationDbContext.CreateExternalEvent is way simpler than before.

HunorTotBagi commented on Modules/Synchronization/src/Synchronization.Infrastructure/Persistence/Database/Configurations/ExternalEventEntityTypeConfiguration.cs:
```diff
@@ -1,3 +1,4 @@
+using Autofac.Builder;
```
Using directive is not required by the code and can be safely removed

HunorTotBagi commented on Modules/Synchronization/src/Synchronization.Infrastructure/Persistence/Database/ValueConverters/ExternalEventPayloadEntityFrameworkValueConverter.cs:
```diff
@@ -3,9 +3,9 @@
+    public ExternalEventPayloadEntityFrameworkValueConverter() : base(
         o => JsonConvert.SerializeObject(o),
         s => JsonConvert.DeserializeObject<ExpandoObject>(s)!)
     { }
```
```suggestion
    {
    }
```

HunorTotBagi commented on Modules/Synchronization/test/Synchronization.Domain.Tests/ExternalEventTests.cs:
For some reason I can't highlight line 4, but here is an unused using directive `using Backbone.UnitTestTools.FluentAssertions.Extensions;`

HunorTotBagi reviewed (changes requested):
I noticed that in `DomainEvents.Incoming` most of the files we have a `Handle` method and a `CreateExternalEvents` but in 

- `IdentityDeletionProcessStartedDomainEventHandler`
- `IdentityDeletionProcessStatusChangedDomainEventHandler`
- `RelationshipStatusChangedDomainEventHandler`

we don't follow this "pattern". So for consistency sake I would refactor the above three mentioned files to have a `Handle` and a `CreateExternalEvents` method. What do you think about this?

ericbrunner commented on Modules/Synchronization/src/Synchronization.Application/DomainEvents/Incoming/PeerDeleted/PeerDeletedDomainEventHandler.cs:
```diff
@@ -25,16 +26,13 @@ public async Task Handle(PeerDeletedDomainEvent domainEvent)
         try
         {
-            await _dbContext.CreateExternalEvent(@event.PeerOfDeletedIdentity, ExternalEventType.PeerDeleted, payload);
+            var payload = new PeerDeletedExternalEvent.PayloadT { RelationshipId = @event.RelationshipId, DeletionDate = SystemTime.UtcNow };
```
I would name the `PayloadT` class either `Payload` or `EventPayload` or has the `T` suffix a special meaning?

tnotheis replied:
The reason for the T was that `Payload` is an inner class of a class that has a `Payload` property. And you cannot have a class and a property with the same name in the same class.

But I like `EventPayload`. Will rename it.

tnotheis commented on Modules/Synchronization/src/Synchronization.Application/Infrastructure/ISynchronizationDbContext.cs:
```diff
@@ -13,7 +13,7 @@ Task<DbPaginationResult<DatawalletModification>> GetDatawalletModifications(Iden
     Task<Datawallet?> GetDatawalletForInsertion(IdentityAddress owner, CancellationToken cancellationToken);
     Task<Datawallet?> GetDatawallet(IdentityAddress owner, CancellationToken cancellationToken);
-    Task<ExternalEvent> CreateExternalEvent(IdentityAddress owner, ExternalEventType type, object payload);
+    Task CreateExternalEvent(ExternalEvent externalEvent);
```
I removed some of the newlines. I only kept the two that were too long for a single line.

ericbrunner commented on Modules/Synchronization/src/Synchronization.Infrastructure/Persistence/Database/SynchronizationDbContext.cs:
```diff
@@ -94,24 +94,20 @@ public async Task<DbPaginationResult<DatawalletModification>> GetDatawalletModif
         await RunInTransaction(async () =>
         {
-            if (externalEvent != null)
-                // if the transaction is retried, the old event has to be removed from the DbSet, because a new one with a new index is added
```
Even with a single statement, surrounding that statement with curly braces enhances readability and even debugging

tnotheis replied:
You commented on removed code. :) The new code doesn't contain this problem anymore.

HunorTotBagi commented:
> I noticed that in `DomainEvents.Incoming` most of the files we have a `Handle` method and a `CreateExternalEvents` but in
> 
> * `IdentityDeletionProcessStartedDomainEventHandler`
> * `IdentityDeletionProcessStatusChangedDomainEventHandler`
> * `RelationshipStatusChangedDomainEventHandler`
> 
> we don't follow this "pattern". So for consistency sake I would refactor the above three mentioned files to have a `Handle` and a `CreateExternalEvents` method. What do you think about this?

@tnotheis Thoughts on this? It seems you missed this one.

tnotheis commented:
@HunorTotBagi Sorry, you're right. I didn't see that.

I just fixed it. I also renamed the methods so they contain the name of the event that is created. So if there is another external event created by the same domain event handler, we would do this in another private method.

Further, I moved the try/catch block to the `Handle` method, so that we don't have to catch exceptions in all of the private methods.

Btw: I think we could even remove the try/catch blocks, because the EventBus classes already log descriptive error messages. But I will not remove them in this PR.

HunorTotBagi commented:
> @HunorTotBagi Sorry, you're right. I didn't see that.
> 
> I just fixed it. I also renamed the methods so they contain the name of the event that is created. So if there is another external event created by the same domain event handler, we would do this in another private method.
> 
> Further, I moved the try/catch block to the `Handle` method, so that we don't have to catch exceptions in all of the private methods.
> 
> Btw: I think we could even remove the try/catch blocks, because the EventBus classes already log descriptive error messages. But I will not remove them in this PR.

Thanks, I like the changes.

I also noticed that the naming of the parameters in the `Handle` and in the private methods are all different for example we can see `domainEvent`, `@event` and `integrationEvent`. We should choose in what method which naming to use and apply all over them.

tnotheis commented:
> I also noticed that the naming of the parameters in the `Handle` and in the private methods are all different for example we can see `domainEvent`, `@event` and `integrationEvent`. We should choose in what method which naming to use and apply all over them.

Done
